In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    certifi-2019.11.28         |           py36_0         149 KB  conda-forge
    ca-certificates-2019.11.28 |       hecc5488_0         145 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1.20.0-py_0       conda-forge

The following packages will be UPDATED:

    ca-

## Extracting the Data

In [2]:
!wget -q -O 'List_of_postal_codes_of_Canada' https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
print('Data downloaded!')

Data downloaded!


In [3]:
data=pd.read_html('List_of_postal_codes_of_Canada')
data[0]

Postcode           Borough  \
0        M1A      Not assigned   
1        M2A      Not assigned   
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A  Downtown Toronto   
8        M8A      Not assigned   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
12       M2B      Not assigned   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
19       M7B      Not assigned   
20       M8B      Not assigned   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M2C      Not assigned   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
35       M7C      Not assigned   
36       M8C      Not assigned   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M2E      Not assigned   
45       M3E      Not assigned   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
49       M7E      Not assigned   
50       M8E      Not assigned   
51       M9E      Not assigned   
52       M1G       Scarborough   
53       M2G      Not assigned   
54       M3G      Not assigned   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
58       M7G      Not assigned   
59       M8G      Not assigned   
60       M9G      Not assigned   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
72       M7H      Not assigned   
73       M8H      Not assigned   
74       M9H      Not assigned   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
87       M7J      Not assigned   
88       M8J      Not assigned   
89       M9J      Not assigned   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M7K      Not assigned   
104      M8K      Not assigned   
105      M9K      Not assigned   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North 

## Dropping rows where the Borough is unassigned 

In [4]:
Toronto_data=data[0]
Toronto_data=Toronto_data[Toronto_data['Borough'] != 'Not assigned']
Toronto_data

Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A  Downtown Toronto   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
52       M1G       Scarborough   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North York   
117      M6L        North York   
118      M6L        North York   
121      M9L        North York   
122      M1M       Scarborough   
123      M1M       Scarborough   
124      M1M       Scarborough   
125      M2M        North York   
126      M2M        North York   
127      M3M        North York   
128      M4M      East Toronto   
129      M5M        North York   
130      M5M        North York   
131      M6M              York   
132      M6M              York   
133      M6M              York   
134      M6M              York   
137      M9M        North York   
138      M9M        North York   
139      M1N       Scarborough   
140      M1N       Scarborough   
141      M2N        North York   
142      M3N        North York   
143      M4N   Central Toronto   
144      M5N   Central Toronto   
145      M6N              York   
146      M6N              York   
149      M9N              York   
150      M1P       Scarbor

In [5]:
Toronto_data.loc[Toronto_data['Neighbourhood'] == 'Not assigned']

Postcode       Borough Neighbourhood
9      M9A  Queen's Park  Not assigned

In [6]:
Toronto_data.loc[Toronto_data['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = 'Queens park'

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/indexing.py:190: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)
/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [7]:
Toronto_data

Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M6A        North York   
6        M6A        North York   
7        M7A  Downtown Toronto   
9        M9A      Queen's Park   
10       M1B       Scarborough   
11       M1B       Scarborough   
13       M3B        North York   
14       M4B         East York   
15       M4B         East York   
16       M5B  Downtown Toronto   
17       M5B  Downtown Toronto   
18       M6B        North York   
21       M9B         Etobicoke   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M1C       Scarborough   
27       M1C       Scarborough   
28       M1C       Scarborough   
30       M3C        North York   
31       M3C        North York   
32       M4C         East York   
33       M5C  Downtown Toronto   
34       M6C              York   
37       M9C         Etobicoke   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M1E       Scarborough   
42       M1E       Scarborough   
43       M1E       Scarborough   
46       M4E      East Toronto   
47       M5E  Downtown Toronto   
48       M6E              York   
52       M1G       Scarborough   
55       M4G         East York   
56       M5G  Downtown Toronto   
57       M6G  Downtown Toronto   
61       M1H       Scarborough   
62       M2H        North York   
63       M3H        North York   
64       M3H        North York   
65       M3H        North York   
66       M4H         East York   
67       M5H  Downtown Toronto   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M6H      West Toronto   
71       M6H      West Toronto   
75       M1J       Scarborough   
76       M2J        North York   
77       M2J        North York   
78       M2J        North York   
79       M3J        North York   
80       M3J        North York   
81       M4J         East York   
82       M5J  Downtown Toronto   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M6J      West Toronto   
86       M6J      West Toronto   
90       M1K       Scarborough   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M2K        North York   
94       M3K        North York   
95       M3K        North York   
96       M4K      East Toronto   
97       M4K      East Toronto   
98       M5K  Downtown Toronto   
99       M5K  Downtown Toronto   
100      M6K      West Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
106      M1L       Scarborough   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M2L        North York   
110      M2L        North York   
111      M3L        North York   
112      M4L      East Toronto   
113      M4L      East Toronto   
114      M5L  Downtown Toronto   
115      M5L  Downtown Toronto   
116      M6L        North York   
117      M6L        North York   
118      M6L        North York   
121      M9L        North York   
122      M1M       Scarborough   
123      M1M       Scarborough   
124      M1M       Scarborough   
125      M2M        North York   
126      M2M        North York   
127      M3M        North York   
128      M4M      East Toronto   
129      M5M        North York   
130      M5M        North York   
131      M6M              York   
132      M6M              York   
133      M6M              York   
134      M6M              York   
137      M9M        North York   
138      M9M        North York   
139      M1N       Scarborough   
140      M1N       Scarborough   
141      M2N        North York   
142      M3N        North York   
143      M4N   Central Toronto   
144      M5N   Central Toronto   
145      M6N              York   
146      M6N              York   
149      M9N              York   
150      M1P       Scarbor

## Now we will combine rows with the same postcode

In [8]:
df = Toronto_data.groupby(['Postcode', 'Borough'], sort = False).agg( ','.join)
df = df.reset_index()
df.head()

# need borough in so it doesn't get written twice

Postcode           Borough                    Neighbourhood
0      M3A        North York                        Parkwoods
1      M4A        North York                 Victoria Village
2      M5A  Downtown Toronto                     Harbourfront
3      M6A        North York  Lawrence Heights,Lawrence Manor
4      M7A  Downtown Toronto                     Queen's Park

In [9]:
df.shape

(103, 3)

## Shape of dataframe is now 103 rows and 3 columns 

## Importing Coordinates for Toronto data

In [10]:
!wget -q -O 'Geospatial coordinates.csv'  http://cocl.us/Geospatial_data
df1 = pd.read_csv('Geospatial coordinates.csv')
df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [11]:
df1=df1.rename(columns = {'Postal Code' : 'Postcode'})
df1

Postcode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476
5        M1J  43.744734 -79.239476
6        M1K  43.727929 -79.262029
7        M1L  43.711112 -79.284577
8        M1M  43.716316 -79.239476
9        M1N  43.692657 -79.264848
10       M1P  43.757410 -79.273304
11       M1R  43.750072 -79.295849
12       M1S  43.794200 -79.262029
13       M1T  43.781638 -79.304302
14       M1V  43.815252 -79.284577
15       M1W  43.799525 -79.318389
16       M1X  43.836125 -79.205636
17       M2H  43.803762 -79.363452
18       M2J  43.778517 -79.346556
19       M2K  43.786947 -79.385975
20       M2L  43.757490 -79.374714
21       M2M  43.789053 -79.408493
22       M2N  43.770120 -79.408493
23       M2P  43.752758 -79.400049
24       M2R  43.782736 -79.442259
25       M3A  43.753259 -79.329656
26       M3B  43.745906 -79.352188
27       M3C  43.725900 -79.340923
28       M3H  43.754328 -79.442259
29       M3J  43.767980 -79.487262
30       M3K  43.737473 -79.464763
31       M3L  43.739015 -79.506944
32       M3M  43.728496 -79.495697
33       M3N  43.761631 -79.520999
34       M4A  43.725882 -79.315572
35       M4B  43.706397 -79.309937
36       M4C  43.695344 -79.318389
37       M4E  43.676357 -79.293031
38       M4G  43.709060 -79.363452
39       M4H  43.705369 -79.349372
40       M4J  43.685347 -79.338106
41       M4K  43.679557 -79.352188
42       M4L  43.668999 -79.315572
43       M4M  43.659526 -79.340923
44       M4N  43.728020 -79.388790
45       M4P  43.712751 -79.390197
46       M4R  43.715383 -79.405678
47       M4S  43.704324 -79.388790
48       M4T  43.689574 -79.383160
49       M4V  43.686412 -79.400049
50       M4W  43.679563 -79.377529
51       M4X  43.667967 -79.367675
52       M4Y  43.665860 -79.383160
53       M5A  43.654260 -79.360636
54       M5B  43.657162 -79.378937
55       M5C  43.651494 -79.375418
56       M5E  43.644771 -79.373306
57       M5G  43.657952 -79.387383
58       M5H  43.650571 -79.384568
59       M5J  43.640816 -79.381752
60       M5K  43.647177 -79.381576
61       M5L  43.648198 -79.379817
62       M5M  43.733283 -79.419750
63       M5N  43.711695 -79.416936
64       M5P  43.696948 -79.411307
65       M5R  43.672710 -79.405678
66       M5S  43.662696 -79.400049
67       M5T  43.653206 -79.400049
68       M5V  43.628947 -79.394420
69       M5W  43.646435 -79.374846
70       M5X  43.648429 -79.382280
71       M6A  43.718518 -79.464763
72       M6B  43.709577 -79.445073
73       M6C  43.693781 -79.428191
74       M6E  43.689026 -79.453512
75       M6G  43.669542 -79.422564
76       M6H  43.669005 -79.442259
77       M6J  43.647927 -79.419750
78       M6K  43.636847 -79.428191
79       M6L  43.713756 -79.490074
80       M6M  43.691116 -79.476013
81       M6N  43.673185 -79.487262
82       M6P  43.661608 -79.464763
83       M6R  43.648960 -79.456325
84       M6S  43.651571 -79.484450
85       M7A  43.662301 -79.389494
86       M7R  43.636966 -79.615819
87       M7Y  43.662744 -79.321558
88       M8V  43.605647 -79.501321
89       M8W  43.602414 -79.543484
90       M8X  43.653654 -79.506944
91       M8Y  43.636258 -79.498509
92       M8Z  43.628841 -79.520999
93       M9A  43.667856 -79.532242
94       M9B  43.650943 -79.554724
95       M9C  43.643515 -79.577201
96       M9L  43.756303 -79.565963
97       M9M  43.724766 -79.532242
98       M9N  43.706876 -79.518188
99       M9P  43.696319 -79.532242
100      M9R  43.688905 -79.554724
101      M9V  43.739416 -79.588437
102      M9W  43.706748 -79.594054

In [13]:
full_data=pd.merge(df, df1, on = 'Postcode')
full_data.head()

Postcode           Borough                    Neighbourhood   Latitude  \
0      M3A        North York                        Parkwoods  43.753259   
1      M4A        North York                 Victoria Village  43.725882   
2      M5A  Downtown Toronto                     Harbourfront  43.654260   
3      M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4      M7A  Downtown Toronto                     Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [16]:

map_Toronto = folium.Map(location=[43.753259, -79.329656], zoom_start=11)

# add markers to map
for lat, lng, label in zip(full_data['Latitude'], full_data['Longitude'], full_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

## Cluster the Neighbourhoods

In [18]:
kclusters = 5

toronto_clus = full_data.drop(['Postcode', 'Borough', 'Neighbourhood'], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_clus)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 2, 3, 2, 1, 0, 4, 4, 2], dtype=int32)

In [19]:
full_data.insert(0, 'Cluster Labels', kmeans.labels_)

In [20]:
full_data

Cluster Labels Postcode           Borough  \
0                 4      M3A        North York   
1                 4      M4A        North York   
2                 2      M5A  Downtown Toronto   
3                 3      M6A        North York   
4                 2      M7A  Downtown Toronto   
5                 1      M9A      Queen's Park   
6                 0      M1B       Scarborough   
7                 4      M3B        North York   
8                 4      M4B         East York   
9                 2      M5B  Downtown Toronto   
10                3      M6B        North York   
11                1      M9B         Etobicoke   
12                0      M1C       Scarborough   
13                4      M3C        North York   
14                4      M4C         East York   
15                2      M5C  Downtown Toronto   
16                2      M6C              York   
17                1      M9C         Etobicoke   
18                0      M1E       Scarborough   
19                4      M4E      East Toronto   
20                2      M5E  Downtown Toronto   
21                2      M6E              York   
22                0      M1G       Scarborough   
23                4      M4G         East York   
24                2      M5G  Downtown Toronto   
25                2      M6G  Downtown Toronto   
26                0      M1H       Scarborough   
27                3      M2H        North York   
28                3      M3H        North York   
29                4      M4H         East York   
30                2      M5H  Downtown Toronto   
31                2      M6H      West Toronto   
32                0      M1J       Scarborough   
33                3      M2J        North York   
34                3      M3J        North York   
35                4      M4J         East York   
36                2      M5J  Downtown Toronto   
37                2      M6J      West Toronto   
38                0      M1K       Scarborough   
39                3      M2K        North York   
40                3      M3K        North York   
41                4      M4K      East Toronto   
42                2      M5K  Downtown Toronto   
43                2      M6K      West Toronto   
44                4      M1L       Scarborough   
45                3      M2L        North York   
46                1      M3L        North York   
47                4      M4L      East Toronto   
48                2      M5L  Downtown Toronto   
49                1      M6L        North York   
50                1      M9L        North York   
51                0      M1M       Scarborough   
52                3      M2M        North York   
53                1      M3M        North York   
54                4      M4M      East Toronto   
55                3      M5M        North York   
56                1      M6M              York   
57                1      M9M        North York   
58                4      M1N       Scarborough   
59                3      M2N        North York   
60                1      M3N        North York   
61                3      M4N   Central Toronto   
62                3      M5N   Central Toronto   
63                1      M6N              York   
64                1      M9N              York   
65                0      M1P       Scarborough   
66                3      M2P        North York   
67                3      M4P   Central Toronto   
68                2      M5P   Central Toronto   
69                1      M6P      West Toronto   
70                1      M9P         Etobicoke   
71                0      M1R       Scarborough   
72                3      M2R        North York   
73                3      M4R   Central Toronto   
74                2      M5R   Central Toronto   
75                2      M6R      West Toronto   
76                1      M7R       Mississauga   
77                1      M9R         Etobicoke   
78                0      M1S       Scarborough   
79   

In [24]:
map_clusters = folium.Map(location=[43.651070,-79.347015],zoom_start=11)

x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, neighbourhood, cluster in zip(full_data['Latitude'], full_data['Longitude'], full_data['Neighbourhood'], full_data['Cluster Labels']):
    label = folium.Popup(' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters